src/DynamoCore/Models/DynamoModel.cs

```diff
@@ -1168,32 +1168,70 @@ private void RegisterHomeWorkspace(HomeWorkspaceModel newWorkspace)
+                // whether some workspaces have already been backed up. If so, those workspaces won't be
+                // backed up again.
+                var tempDict = new Dictionary<Guid,string>(backupFilesDict);
+                backupFilesDict.Clear();
+                PreferenceSettings.BackupFiles.Clear();
                 foreach (var workspace in Workspaces)
                 {
                     if (!workspace.HasUnsavedChanges)
-                        continue;
-
-                    string fileName;
-                    if (string.IsNullOrEmpty(workspace.FileName))
                     {
-                        fileName = Configurations.BackupFileNamePrefix + workspace.Guid;
-                        var ext = workspace is HomeWorkspaceModel ? ".DYN" : ".DYF";
-                        fileName += ext;
-                    }
-                    else
-                    {
-                        fileName = Path.GetFileName(workspace.FileName);
+                        if (workspace.Nodes.Count == 0 &&
+                            workspace.Notes.Count == 0)
+                            continue;
+
+                        if (tempDict.ContainsKey(workspace.Guid))
+                        {
+                            backupFilesDict.Add(workspace.Guid, tempDict[workspace.Guid]);
+                            continue;
+                        }
                     }
 
-                    var savePath = Path.Combine(pathManager.BackupDirectory, fileName);
+                    var savePath = GetBackupFilePath(workspace);
+                    var oldFileName = workspace.FileName;
+                    var oldName = workspace.Name;
                     workspace.SaveAs(savePath, null);
+                    workspace.FileName = oldFileName;
+                    workspace.Name = oldName;
+                    backupFilesDict.Add(workspace.Guid, savePath);
                     Logger.Log("Backup file is saved: " + savePath);
                 }
+                PreferenceSettings.BackupFiles.AddRange(backupFilesDict.Values);
             });
         }
 
         /// <summary>
+        /// Get the backup file path for a workspace
+        /// </summary>
+        /// <param name="workspace"></param>
+        /// <returns></returns>
+        private string GetBackupFilePath(WorkspaceModel workspace)
+        {
+            string fileName;
+            if (string.IsNullOrEmpty(workspace.FileName))
+            {
+                if (workspace is HomeWorkspaceModel)
+                {
+                    fileName = Configurations.BackupFileNamePrefix + ".DYN";
+                }
+                else
+                {
+                    fileName = workspace.Name + ".DYF";
+                }
+            }
+            else
+            {
+                fileName = Path.GetFileName(workspace.FileName);
+            }
+
+            return Path.Combine(pathManager.BackupDirectory, fileName);
+        }
+
+        /// <summary>
```

Comment:
This is a *path translation* method and it shouldn't be here, please move it to `PathManager` class (doesn't have to be on the `IPathManager` interface) as the following method:

```csharp
class PathManager : IPathManager
{
    internal string GetBackupFilePath(WorkspaceModel workspace);
}
```